Handle funtemp redecl chain

martong opened the pull request.

Commit:
Add structural eq tests for template args

Commit:
Lookup finds previous friend function template

martong commented on unittests/AST/ASTImporterTest.cpp:
```diff
@@ -1996,6 +1996,193 @@ TEST_P(ASTImporterOptionSpecificTestBase, ReimportWithUsedFlag) {
   EXPECT_TRUE(Imported2->isUsed(false));
 }
 
+// FIXME Can we unify these tests with ImportClassTemplates?
```
I could unify these tests similarly to the Visibility tests in #502 , so they would be both type and value parameterized tests.  In one hand we will not have that much repetition of the test code, but on the other hand understanding it will be more difficult. @balazske what do you think?

balazske replied:
I think in case of `ASTImporter` there are many places for test generalization because many similar test cases, so it is better to have general tests that work with every possible parameter combination. Otherwise we can have too many test cases.

martong commented:
run tests

gamesh411 commented on unittests/AST/ASTImporterTest.cpp:
```diff
@@ -3692,6 +3692,158 @@ TEST_P(ImportClasses, ImportPrototypeThenDefinition) {
+  ASSERT_NE(D, D1);
+  auto Res = lookup(D1->getDeclName());
+  // Ptr
+  dump(Res);
```
Debugging dump statement left in.

gamesh411 commented on unittests/AST/ASTImporterTest.cpp:
```diff
@@ -3692,6 +3692,158 @@ TEST_P(ImportClasses, ImportPrototypeThenDefinition) {
+  // Vector
+  auto Res = lookup(D1->getDeclName());
+  EXPECT_EQ(Res.size(), 3u);
+  dump(Res);
```
Debugging dump statement left in.

balazske commented on lib/AST/ASTImporter.cpp:
```diff
@@ -5016,27 +5072,32 @@ Decl *ASTNodeImporter::VisitFunctionTemplateDecl(FunctionTemplateDecl *D) {
-            // FIXME: Actually try to merge the body and other attributes.
-            return FoundFunction;
+          if (IsStructuralMatch(D, FoundTemplate)) {
+            FunctionTemplateDecl* TemplateWithDef = getDefinition(FoundTemplate);
```
`*` is at wrong place

balazske commented on unittests/AST/ASTImporterTest.cpp:
```diff
@@ -1996,6 +1996,193 @@ TEST_P(ASTImporterOptionSpecificTestBase, ReimportWithUsedFlag) {
+       PrototypeShouldBeImportedAsAPrototypeWhenThereIsNoDefinition) {
+  Decl *FromTU = getTuDecl("template <class T> void X();", Lang_CXX);
+  auto Pattern = functionTemplateDecl(hasName("X"), unless(isImplicit()));
+  auto FromD = FirstDeclMatcher<FunctionTemplateDecl>().match(FromTU, Pattern);
```
I do not know why but `auto *` is the usual way of this kind of declarations (in this file).

Commit:
Unify redecl chain tests as type parameterized tests

Commit:
Handle redecl chain of FunctionTemplateDecls

martong commented:
Update: 
- Removed the obsolate `idBasedLookup` since we use the astimporter specific lookup since #505 
- Rebased on top of #508, so please check the last 3 commits.

martong commented:
https://reviews.llvm.org/D57235
